here-naksha-lib-psql/src/commonMain/kotlin/naksha/psql/executors/query/WhereClauseBuilder.kt

```diff
@@ -0,0 +1,221 @@
+                )
+                val placeholder = placeholderForArg(twkb, PgType.BYTE_ARRAY)
+                where.append("ST_Intersects(naksha_geometry(${PgColumn.geo}, ${PgColumn.flags}), ST_GeomFromTWKB($placeholder))")
+            }
+
+            else -> throw NakshaException(
+                NakshaError.ILLEGAL_ARGUMENT,
+                "Invalid spatial query found: $spatial"
+            )
+        }
+    }
+
+    private fun whereMetadata() {
+        request.query.metadata?.let { metaQuery ->
+            if (where.isNotEmpty()) {
+                where.append(" AND ")
+            }
+            where.append("(")
+            whereNestedMetadata(request, metaQuery, where)
+            where.append(")")
+        }
+    }
+
+    private tailrec fun whereNestedMetadata(
+        request: ReadFeatures,
+        metaQuery: IMetaQuery,
+        where: StringBuilder
+    ) {
+        when (metaQuery) {
+            is MetaNot -> {
+                where.append(" NOT ")
+                whereNestedMetadata(request, metaQuery.query, where)
+            }
+
+            is MetaAnd -> {
+                metaQuery.filterNotNull().forEachIndexed { index, subQuery ->
+                    if (index > 0) {
+                        where.append(" AND ")
+                    }
+                    whereNestedMetadata(request, subQuery, where)
+                }
+            }
+
+            is MetaOr -> {
+                metaQuery.filterNotNull().forEachIndexed { index, subQuery ->
+                    if (index > 0) {
+                        where.append(" OR ")
+                    }
+                    whereNestedMetadata(request, subQuery, where)
+                }
+            }
+
+            is MetaQuery -> {
+                val pgColumn = PgColumn.ofRowColumn(metaQuery.column) ?: throw NakshaException(
+                    NakshaError.ILLEGAL_STATE,
+                    "Couldn't find PgColumn for TupleColumn: ${metaQuery.column.name}"
+                )
+                val placeholder = placeholderForArg(metaQuery.value, pgColumn.type)
+                val resolvedQuery = when (val op = metaQuery.op) {
+                    is StringOp -> resolveStringOp(op, pgColumn, placeholder)
+                    is DoubleOp -> resolveDoubleOp(op, pgColumn, placeholder)
+                    else -> throw NakshaException(
+                        NakshaError.ILLEGAL_ARGUMENT,
+                        "Unknown op type: ${op::class.simpleName}"
+                    )
+                }
+                where.append(resolvedQuery)
+            }
+
+            else -> throw NakshaException(
+                NakshaError.ILLEGAL_ARGUMENT,
+                "Unknown metadata query type: ${metaQuery::class.simpleName}"
+            )
+        }
+    }
+
+    private fun placeholderForArg(value: Any?, type: PgType): String {
+        argValues.add(value)
+        argTypes.add(type.toString())
+        return "\$${argTypes.size}"
```

Comment:
Shouldn't this be returning argValues.toString() ? In `whereNestedMetadata()` you seems to use this as input for `resolveStringOp()` and `resolveDoubleOp()`, which in turn is `valuePlaceHolder`.

Reply:
Not really, let me explain what happens in this method - but first let's take a look at the documentation of `PgPlan.execute` method:

```
/**
 * Execute the prepared plan with the given arguments. The types must match to the prepared statement.
 * @param args the arguments to be set at $n position, where $1 is the first array element.
 * @return either the number of affected rows or the rows.
 */
fun execute(args: Array<Any?>? = null): PgCursor
```

This sentence is crucial: "arguments to be set at $n position, where $1 is the first array element".
So basically, in the current postgres API we are using, when we want to execute some **parameterized** query on DB, we must provide 3 things:
- SQL statement
- argument values
- argument types

For example, when we want to get features that have specific type and where created after some times, the query would be something like (its naive but let's stick for simplicity):
```
SELECT * FROM features_table WHERE type = $1 AND createdAt >= $2
```

Before we pass this thing to Postgres, we must take care of the arguments ($1, $2) - the driver under the hood needs types and values (so let's say `"topology"` which is `string` for `$1` and `123456789` which is `int` for `$2`)

... now coming back to our builder and your comment ;) 
The `placeholderForArg` does 3 things that need to be always done together when we add some argument to the query:
- keep track of the value (`argValues.add(value)`)
- keep track of value's type (`argTypes.add(type.toString())`)
- return the `$N` to be placed in the sql query (which later will be resolved)

Now the important thing to notice is that indices of all these three things (values, types, placeholder) must match. Trick is, placeholder starts with 1, but this is minor.

So, glueing it together with our example above:
```
val topologyPlaceholder = placeholderFor("topology", PgString) // returns "$1"
val createdAtPlaceholder = placeholderFor(123456789, PgInt) // returns "$2"
val ourWhereClause = "type = $topologyPlaceholder AND createdAt >= $createdAtPlaceholder"

where.append(ourWhereClause) 

doStuffWith(where, argValues, argTypes) // it knows everything it needs
```

Reply:
Understood, this is still before the args are injected into the query. Thanks for your very detailed explanation 👍